packages/node-core/src/indexer/dictionary/types.ts

```diff
@@ -0,0 +1,46 @@
+// Copyright 2020-2023 SubQuery Pte Ltd authors & contributors
+// SPDX-License-Identifier: GPL-3.0
+
+import {DictionaryQueryEntry as DictionaryV1QueryEntry} from '@subql/types-core/dist/project/types';
+import {MetaData as DictionaryV1Metadata} from '@subql/utils';
+import {BlockHeightMap} from '../../utils/blockHeightMap';
+import {DictionaryV2Metadata, DictionaryV2QueryEntry} from './';
+
+export type DictionaryResponse<B> = {
+  batchBlocks: number[] | B[];
+  lastBufferedHeight: number;
+};
+
+export enum DictionaryVersion {
+  v1 = 'v1',
+  v2Basic = 'v2Basic',
+  v2Complete = 'v2Complete',
+}
+
+export interface IDictionary<DS, FB> {
+  getData(
+    startBlock: number,
+    queryEndBlock: number,
+    limit: number
+  ): Promise<DictionaryResponse<FB | number> | undefined>;
+  initMetadata(): Promise<void>;
+  queryMapValidByHeight(height: number): boolean;
+  getQueryEndBlock(startHeight: number, apiFinalizedHeight: number): number;
+  version: DictionaryVersion;
+  startHeight: number;
+  heightValidation(height: number): boolean;
+  updateQueriesMap(dataSources: BlockHeightMap<DS[]>): void;
+}
+
+export interface IDictionaryCtrl<DS, FB, D extends IDictionary<DS, FB>> {
+  initDictionaries(apiGenesisHash: string): Promise<void>;
+  dictionary: D;
+  useDictionary: boolean;
+  findDictionary(height: number): Promise<void>;
```

Comment:
I think this can be private, `scopedDictionaryEntries` can call this.

Reply:
we  also call `findDictionary ` in fetch.service at init `fillNextBlockBuffer`. Then we are able to check `useDictionary`, if not `scopedDictionaryEntries ` will never been called and fetch will use original way to enqueue blocks